File: Python/Linked List/reverseLL.py
# class node:
#     def __init__(self,data):
#         self.data = data
#         self.next = None

# a = node(10)
# b = node(20)
# a.next = b
# print(a.data)
# print(b.data)
# print(a.next.data)

class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

def printLL(head):
    while head is not None:
        print(str(head.data) + "->",end="")
        head = head.next
    print("None")
    return

def insert(head,pos,data):
    i = 0
    curr = head
    while(i!=pos):
        head = curr
        curr = curr.next
        i += 1
    new_node = Node(data)
    new_node.next = curr
    head.next = new_node

def Reverse(head):
    Next = head
    Curr = None
    Prev = None
    while Next is not None:
        Prev = Curr
        Curr = Next
        Next = Next.next
        Curr.next = Prev
    head = Curr
    return head


def takeInput():
    # inputList = [int(ele) for ele in input().split()]
    inputList = [1,2,3,4,5]
    head = None
    tail = None
    for currData in inputList:
        if currData == -1:
            break

        new_node = Node(currData)
        if head is None:
            head = new_node
            tail = new_node
        else:
            tail.next = new_node
            tail = new_node
    return head

head = takeInput()
insert(head,2,6)
printLL(head)
head = Reverse(head)
printLL(head)

File: .vscode/Linked List/reverse_LL.cpp
#include<iostream>
using namespace std;

class Node {
    public:

        int data;
        Node *next;
};

void create(Node **h,int d){
    Node *new_node = new Node;
    Node *last = *h;

    new_node->data = d;
    new_node->next = NULL;

    if(*h==NULL){
        *h=new_node;
        return;
    }

    while(last->next != NULL){
        last = last->next;
    }
    last->next = new_node;
    return;
}

void reverse(Node **h,Node *p,Node *q,Node *r){

    while(p!=NULL){
        r=q;
        q=p;
        p=p->next;
        q->next=r;
    }
    *h=q;

}
void recursiveReverse(Node **h,Node *p,Node *q){
    if(p!=NULL){
        recursiveReverse(h,p->next,p);
        p->next=q;
    }else{
        *h=q;
    }
}

void display(Node *h){
    while(h!=NULL){
        cout<<h->data<<endl;
        h=h->next;
    }
}


int main(){
    Node *head = NULL;
    
    create(&head,10);
    create(&head,20);
    create(&head,30);
    create(&head,40);
    create(&head,50);
    create(&head,60);
    create(&head,70);create(&head,80);create(&head,90);create(&head,100);
    display(head);
    Node *p=head;
    Node *q=NULL;
    Node *r=NULL;
    //reverse(&head,p,q,r);
    recursiveReverse(&head,p,q);
    display(head);

}